agent/agent.go

```diff
@@ -62,13 +66,87 @@ var (
 	healthCheckPath      = flag.String("health-check-path", "/", "Path on backend host to issue health checks against.  Defaults to the root.")
 	healthCheckFreq      = flag.Int("health-check-interval-seconds", 0, "Wait time in seconds between health checks.  Set to zero to disable health checks.  Checks disabled by default.")
 	healthCheckUnhealthy = flag.Int("health-check-unhealthy-threshold", 2, "A so-far healthy backend will be marked unhealthy after this many consecutive failures. The minimum value is 1.")
+	sessionCookieTimeout = flag.Duration("session-cookie-timeout", 10*time.Minute, "Cookie timeout for the session cookie")
+	cookieLRU, _         = utils.NewCookieCache(cookieCacheLimit)
 )
 
+func remove(slice []*http.Cookie, i int) ([]*http.Cookie, *http.Cookie) {
+	sessionCookie := slice[i]
+	copy(slice[i:], slice[i+1:])
+	return slice[:len(slice)-1], sessionCookie
+}
+
 func hostProxy(ctx context.Context, host, shimPath string, injectShimCode bool) (http.Handler, error) {
 	hostProxy := httputil.NewSingleHostReverseProxy(&url.URL{
 		Scheme: "http",
 		Host:   host,
 	})
+
+	existingDirector := hostProxy.Director
+
+	hostProxy.Director = func(r *http.Request) {
+		existingDirector(r)
+
+		sessionCookie, err := r.Cookie("session-id")
+		// requestDump, err := httputil.DumpRequest(r, true)
+		// if err != nil {
+		// 	log.Println(err)
+		// }
+		// log.Println("request in director: ", string(requestDump))
+
+		if err != nil {
+			// This should not happen, should have a session cookie by this point
+			log.Fatal(err)
+		} else {
+			if sessionCookie != nil {
+				sessionID := sessionCookie.Value
+				cachedCookieJar, ok := cookieLRU.GetCachedCookieJar(sessionID)
+				if !ok {
+					fmt.Printf("No cached cookie jar with session ID: %v", sessionID)
+				}
+
+				// Get only the applicable cookies
+				cookies := cachedCookieJar.(http.CookieJar).Cookies(r.URL)
+				// Add the cookies to the request header
+				for c := range cookies {
+					r.AddCookie(cookies[c])
+				}
+			}
+
+		}
+	}
+
+	hostProxy.ModifyResponse = func(res *http.Response) error {
+
+		//log.Println("header in modifyresponse ", res.Header)
+		// responseDump, err := httputil.DumpResponse(res, true)
+		// if err != nil {
+		// 	log.Println(err)
+		// }
+		// log.Println("response in modify: ", string(responseDump))
+
+		// need to get session cookie first to get the session id
```

Comment:
I would not expect the response from the backend to include the session cookie.

However, it does have the request modified by the Director method above in a field called `Request`.

So, try doing something like `res.Request.Cookies()`
Instead, we should have a cookie